File: app/src/main/java/com/pandatone/kumiwake/kumiwake/KumiwakeCustom.java
package com.pandatone.kumiwake.kumiwake;

import android.content.Intent;
import android.graphics.Color;
import android.graphics.Point;
import android.graphics.Rect;
import android.os.Build;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.text.InputFilter;
import android.transition.Slide;
import android.util.Log;
import android.view.View;
import android.view.ViewTreeObserver;
import android.view.Window;
import android.view.WindowManager;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.ScrollView;
import android.widget.Spinner;
import android.widget.TextView;

import com.pandatone.kumiwake.R;
import com.pandatone.kumiwake.adapter.EditGroupListAdapter;
import com.pandatone.kumiwake.member.GroupListAdapter;
import com.pandatone.kumiwake.adapter.MBListViewAdapter;
import com.pandatone.kumiwake.member.Name;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by atsushi_2 on 2016/05/27.
 */
public class KumiwakeCustom extends AppCompatActivity {

    public static ListView memberList, groupList;
    static TextView numberOfSelectedMember, groupNo;
    Button addMember;
    ImageView backgroundImage;
    Spinner spinner;
    public static ScrollView scrollView;
    CheckBox even_fm_ratio_check, even_age_ratio_check, even_grade_ratio_check;
    MBListViewAdapter mbAdapter;
    EditGroupListAdapter gpAdapter;
    public static ArrayList<Name> memberArray, origMemberArray;
    static ArrayList<GroupListAdapter.Group> groupArray, newGroupArray;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN);
        setContentView(R.layout.kumiwake_custom);
        ButterKnife.bind(this);

        memberArray = (ArrayList<Name>) getIntent().getSerializableExtra("NormalModeMemberArray");
        groupArray = (ArrayList<GroupListAdapter.Group>) getIntent().getSerializableExtra("NormalModeGroupArray");
        mbAdapter = new MBListViewAdapter(this, memberArray, groupArray.size());
        gpAdapter = new EditGroupListAdapter(this, groupArray);
        findViews();
        setViews();
        memberList.setAdapter(mbAdapter);
        groupList.setAdapter(gpAdapter);
        setLeader();

        memberList.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            //行をクリックした時の処理
            public void onItemClick(final AdapterView<?> parent, View view, final int position, long id) {
                memberList.requestFocus();
                changeLeader(position);
            }
        });

        scrollView.post(new Runnable() {
            public void run() {
                scrollView.fullScroll(ScrollView.FOCUS_UP);
            }
        });

        origMemberArray = memberArray;
    }

    @Override
    protected void onStart() {
        super.onStart();

        /*memberArray = origMemberArray;
        mbAdapter = new MBListViewAdapter(this, memberArray, groupArray.size());
        setViews();
        memberList.setAdapter(mbAdapter);
        setLeader();*/


    }

    public void findViews() {
        scrollView = (ScrollView) findViewById(R.id.custom_scroll);
        memberList = (ListView) findViewById(R.id.add_group_listview).findViewById(R.id.reviewListView);
        numberOfSelectedMember = (TextView) findViewById(R.id.add_group_listview).findViewById(R.id.numberOfSelectedMember);
        memberList.setEmptyView(findViewById(R.id.add_group_listview).findViewById(R.id.emptyMemberList));
        addMember = (Button) findViewById(R.id.add_group_listview).findViewById(R.id.member_add_btn);
        groupList = (ListView) findViewById(R.id.kumiwake_group_listView);
        groupNo = (TextView) findViewById(R.id.group_no);
        backgroundImage = (ImageView) findViewById(R.id.background_img);


        even_fm_ratio_check = (CheckBox) findViewById(R.id.even_fm_ratio_check);
        even_age_ratio_check = (CheckBox) findViewById(R.id.even_age_ratio_check);
        even_grade_ratio_check = (CheckBox) findViewById(R.id.even_grade_ratio_check);
        even_age_ratio_check.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (even_age_ratio_check.isChecked() == true) {
                    even_grade_ratio_check.setEnabled(false);
                    even_grade_ratio_check.setChecked(false);
                } else {
                    even_grade_ratio_check.setEnabled(true);
                }
            }
        });
        even_grade_ratio_check.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (even_grade_ratio_check.isChecked() == true) {
                    even_age_ratio_check.setEnabled(false);
                    even_age_ratio_check.setChecked(false);
                } else {
                    even_age_ratio_check.setEnabled(true);
                }
            }
        });

        spinner = (Spinner) findViewById(R.id.custom_spinner);
        ArrayAdapter<String> adapter = new ArrayAdapter<String>(this, android.R.layout.simple_spinner_item);
        adapter.setDropDownViewResource(
                android.R.layout.simple_spinner_dropdown_item
        );
        String[] strs = getResources().getStringArray(R.array.role);
        List<String> list = new ArrayList<String>(Arrays.asList(strs)); // 新インスタンスを生成
        list.remove(1);
        adapter.addAll(list);
        spinner.setAdapter(adapter);
    }

    public void setViews() {
        addMember.setVisibility(View.GONE);
        MBListViewAdapter.setRowHeight(memberList, mbAdapter);
        EditGroupListAdapter.setRowHeight(groupList, gpAdapter);
        numberOfSelectedMember.setText(memberArray.size() + getString(R.string.person));
        groupNo.setText(String.valueOf(groupArray.size()) + " " + getText(R.string.group));
        Point size = new Point();
        getWindowManager().getDefaultDisplay().getSize(size);
        int screenHeight = size.y;
        backgroundImage.getLayoutParams().height = screenHeight;
    }

    public void setLeader() {
        for (int i = 0; i < memberArray.size(); i++) {
            if (memberArray.get(i).getRole()!=null) {
                if (memberArray.get(i).getRole().matches(".*" + getText(R.string.leader) + ".*")) {

                    String[] roleArray = memberArray.get(i).getRole().split(",");
                    List<String> list = new ArrayList<String>(Arrays.asList(roleArray));
                    Set<String> hs = new HashSet<>();
                    hs.addAll(list);
                    list.clear();
                    list.addAll(hs);
                    list.remove("");
                    Collections.sort(list);
                    final int nowLDNo = Integer.parseInt(list.get(0).substring(2));
                    final int finalI = i;
                    groupList.post(new Runnable() {
                        @Override
                        public void run() {
                            TextView leader = (TextView) groupList.getChildAt(nowLDNo - 1).findViewById(R.id.leader);
                            leader.setText(getText(R.string.leader) + ":" + memberArray.get(finalI).getName());
                        }
                    });
                }
            }
        }
    }

    @OnClick(R.id.normal_kumiwake_button)
    void onClicked() {
        int memberSum = 0;
        Boolean allowToNext = true;
        for (int i = 0; i < groupList.getCount(); i++) {
            int memberNo = EditGroupListAdapter.getMemberNo(i);
            memberSum += memberNo;
            if (memberNo <= 0 || memberSum > memberArray.size()) {
                allowToNext = false;
            }
        }

        if (allowToNext) {
            recreateGrouplist();
            Intent intent = new Intent(this, NormalKumiwakeConfirmation.class);
            intent.putExtra("NormalModeMemberArray", memberArray);
            intent.putExtra("NormalModeGroupArray", newGroupArray);
            intent.putExtra("EvenFMRatio", even_fm_ratio_check.isChecked());
            intent.putExtra("EvenAgeRatio", even_age_ratio_check.isChecked());
            intent.putExtra("EvenGradeRatio", even_grade_ratio_check.isChecked());
            intent.putExtra("EvenRole", (String) spinner.getSelectedItem());
            startActivity(intent);
            overridePendingTransition(R.anim.in_right, R.anim.out_left);
        } else {
            TextView error_mbNo = (TextView) findViewById(R.id.error_member_no);
            error_mbNo.setVisibility(View.VISIBLE);
        }
    }

    @OnClick(R.id.back_to_initial_mbNo)
    void onbcClicked() {
        EditText et;
        for (int i = 0; i < groupList.getCount(); i++) {
            et = (EditText) groupList.getChildAt(i).findViewById(R.id.editTheNumberOfMember);
            et.setFocusable(false);
            et.setText(String.valueOf(groupArray.get(i).getBelongNo()));
            et.setFocusableInTouchMode(true);
            et.setTextColor(Color.BLACK);
        }
    }

    public void changeLeader(int position) {
        StringBuilder newRole = new StringBuilder();
        String roleItem = memberArray.get(position).getRole();
        if (roleItem.matches(".*" + getText(R.string.leader) + ".*")) {
            List<String> list = deleteLeaderList(roleItem);
            int nowLDNo = Integer.parseInt(list.get(0).substring(2));
            MBListViewAdapter.LDNoArray.remove(MBListViewAdapter.LDNoArray.indexOf(nowLDNo));
            list.remove(list.get(0));
            TextView leader = (TextView) groupList.getChildAt(nowLDNo - 1).findViewById(R.id.leader);
            leader.setText(getText(R.string.leader) + ":" + getText(R.string.nothing));

            for (int j = 0; j < list.size(); j++) {
                newRole.append(list.get(j));
                if (j != list.size() - 1) {
                    newRole.append(",");
                }
            }
        } else {
            MBListViewAdapter.LDNo = 1;
            while (MBListViewAdapter.LDNoArray.contains(MBListViewAdapter.LDNo)) {
                MBListViewAdapter.LDNo++;
            }
            if (MBListViewAdapter.LDNo != groupArray.size() + 1) {
                newRole.append(roleItem);
                newRole.append("," + getText(R.string.leader));
                TextView leader = (TextView) groupList.getChildAt(MBListViewAdapter.LDNo - 1).findViewById(R.id.leader);
                leader.setText(getText(R.string.leader) + ":" + memberArray.get(position).getName());
            }
        }
        memberArray.set(position, new Name(memberArray.get(position).getId(), memberArray.get(position).getName(),
                memberArray.get(position).getSex(), memberArray.get(position).getAge(),
                memberArray.get(position).getGrade(), memberArray.get(position).getBelong(), newRole.toString(),
                memberArray.get(position).getName_read()));
        mbAdapter.notifyDataSetChanged();
    }

    public static List<String> deleteLeaderList(String roleItem) {
        String[] roleArray = roleItem.split(",");
        List<String> list = new ArrayList<String>(Arrays.asList(roleArray));
        Set<String> hs = new HashSet<>();
        hs.addAll(list);
        list.clear();
        list.addAll(hs);
        list.remove(MainActivity.getContext().getText(R.string.leader));
        list.remove("");
        Collections.sort(list);
        return list;
    }

    public void recreateGrouplist() {
        newGroupArray = new ArrayList<GroupListAdapter.Group>();
        for (int i = 0; i < groupList.getCount(); i++) {
            String groupName = EditGroupListAdapter.getGroupName(i);
            int memberNo = EditGroupListAdapter.getMemberNo(i);
            newGroupArray.add(new GroupListAdapter.Group(i, groupName, memberNo, null));
        }
    }

    public static void changeBelongNo(int position, int addNo) {
        EditText et;
        int nowNo = 0, newNo = 0;
        if (position == groupList.getCount() - 1) {
            et = (EditText) groupList.getChildAt(0).findViewById(R.id.editTheNumberOfMember);
        } else {
            et = (EditText) groupList.getChildAt(position + 1).findViewById(R.id.editTheNumberOfMember);
        }
        if (et.getText().toString().length() > 0) {
            nowNo = Integer.parseInt(et.getText().toString());
        }
        newNo = nowNo + addNo;
        et.setFilters(new InputFilter[]{new InputFilter.LengthFilter(3)});
        et.setText(String.valueOf(newNo));
        if (newNo < 0) {
            et.setTextColor(Color.RED);
        } else {
            et.setTextColor(Color.BLACK);
        }
    }

}

File: app/src/main/java/com/pandatone/kumiwake/kumiwake/KumiwakeCustom.kt
package com.pandatone.kumiwake.kumiwake

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.Color
import android.graphics.Point
import android.graphics.Rect
import android.os.Bundle
import android.view.View
import android.view.ViewTreeObserver
import android.view.WindowManager
import android.widget.AdapterView
import android.widget.Button
import android.widget.EditText
import android.widget.ListView
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.content.ContextCompat
import com.pandatone.kumiwake.KumiwakeArrayKeys
import com.pandatone.kumiwake.KumiwakeCustomKeys
import com.pandatone.kumiwake.ModeKeys
import com.pandatone.kumiwake.R
import com.pandatone.kumiwake.StatusHolder
import com.pandatone.kumiwake.adapter.EditGroupViewAdapter
import com.pandatone.kumiwake.adapter.SmallMBListAdapter
import com.pandatone.kumiwake.databinding.KumiwakeCustomBinding
import com.pandatone.kumiwake.member.function.Group
import com.pandatone.kumiwake.member.function.Member


/**
 * Created by atsushi_2 on 2016/05/27.
 */
class KumiwakeCustom : AppCompatActivity() {
    private lateinit var binding: KumiwakeCustomBinding
    private lateinit var memberListView: ListView
    private lateinit var groupListView: ListView
    private var mbAdapter: SmallMBListAdapter? = null
    private var editGPAdapter: EditGroupViewAdapter? = null
    private lateinit var memberArray: ArrayList<Member>
    private lateinit var groupArray: ArrayList<Group>
    private lateinit var newGroupArray: ArrayList<Group>
    private var leaderArray: ArrayList<Member?> = ArrayList()
    private var screenHeight = 0
    private var nextSet = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_HIDDEN)
        setTheme(StatusHolder.nowTheme)
        binding = KumiwakeCustomBinding.inflate(layoutInflater)
        setContentView(binding.root)
        if (StatusHolder.mode == ModeKeys.Sekigime.key) {
            val layout = findViewById<ConstraintLayout>(R.id.custom_root_layout)
            layout.background = ContextCompat.getDrawable(this, R.drawable.sekigime_background)
        }

        if (intent.getSerializableExtra(KumiwakeArrayKeys.MEMBER_LIST.key) != null) {
            memberArray =
                intent.getSerializableExtra(KumiwakeArrayKeys.MEMBER_LIST.key) as ArrayList<Member>
        }
        if (intent.getSerializableExtra(KumiwakeArrayKeys.GROUP_LIST.key) != null) {
            groupArray =
                intent.getSerializableExtra(KumiwakeArrayKeys.GROUP_LIST.key) as ArrayList<Group>
        }
        findViews()
        mbAdapter = SmallMBListAdapter(this, memberArray, showLeaderNo = true)
        editGPAdapter = EditGroupViewAdapter(this, groupArray, groupListView)
        setViews()
        memberListView.adapter = mbAdapter
        groupListView.adapter = editGPAdapter

        setKeyboardListener()
        memberListView.onItemClickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
            //行をクリックした時の処理
            memberListView.requestFocus()
            changeLeader(position)
        }
        val leaderArrayB: Array<Member?> = arrayOfNulls(groupArray.size) //n番目にグループnのリーダーを格納
        leaderArray = leaderArrayB.toCollection(ArrayList())
    }

    //View宣言
    private fun findViews() {
        memberListView = findViewById(R.id.memberListView)
        groupListView = findViewById(R.id.groupListView)
        memberListView.emptyView = findViewById(R.id.emptyMemberList)
    }

    //View初期化
    @SuppressLint("SetTextI18n")
    fun setViews() {
        binding.addGroupListView.memberAddBtn.visibility = View.GONE
        binding.addGroupListView.memberRegisterAndAddBtn.visibility = View.GONE
        mbAdapter?.setRowHeight(memberListView)
        editGPAdapter?.setRowHeight(groupListView)
        binding.addGroupListView.numberOfSelectedMember.text =
            memberArray.size.toString() + getString(R.string.people)
        binding.groupNoTxt.text = groupArray.size.toString() + " " + getText(R.string.group)
        val size = Point()
        windowManager.defaultDisplay.getSize(size)
        screenHeight = size.y
        findViewById<Button>(R.id.normal_kumiwake_button).setOnClickListener { onNextClicked() }
        findViewById<Button>(R.id.back_to_initial_mbNo).setOnClickListener { onBCClicked() }
    }

    //組み分け確認画面に遷移ボタン
    private fun onNextClicked() {
        var memberSum = 0
        var allowToNext: Boolean? = true
        for (i in 0 until groupListView.count) {
            val memberNo = editGPAdapter!!.getMemberNo(i)
            memberSum += memberNo
            if (memberNo <= 0 || memberSum > memberArray.size) {
                allowToNext = false
            }
        }

        if (allowToNext!!) {
            createGroupArray()
            val intent = Intent(this, KumiwakeConfirmation::class.java)
            intent.putExtra(KumiwakeArrayKeys.MEMBER_LIST.key, memberArray)
            intent.putExtra(KumiwakeArrayKeys.GROUP_LIST.key, newGroupArray)
            intent.putExtra(KumiwakeArrayKeys.LEADER_LIST.key, leaderArray)
            intent.putExtra(
                KumiwakeCustomKeys.EVEN_FM_RATIO.key,
                binding.evenFmRatioCheck.isChecked
            )
            intent.putExtra(
                KumiwakeCustomKeys.EVEN_AGE_RATIO.key,
                binding.evenAgeRatioCheck.isChecked
            )
            startActivity(intent)
            overridePendingTransition(R.anim.in_right, R.anim.out_left)
        } else {
            binding.errorMemberNoTxt.visibility = View.VISIBLE
        }
    }

    //初期状態に戻すボタン
    private fun onBCClicked() {
        var et: EditText
        for (i in 0 until groupListView.count) {
            et = groupListView.getChildAt(i)
                .findViewById<View>(R.id.editTheNumberOfMember) as EditText
            et.isFocusable = false
            et.setText(groupArray[i].belongNo.toString())
            et.isFocusableInTouchMode = true
            et.setTextColor(Color.BLACK)
        }
    }

    //リーダーの選択・解除処理
    @SuppressLint("SetTextI18n")
    fun changeLeader(position: Int) {
        val selectedMember = memberArray[position]
        //リーダー解除
        if (leaderArray.contains(selectedMember)) {
            nextSet = leaderArray.indexOf(selectedMember)
            leaderArray[nextSet] = null
            val leader =
                groupListView.getChildAt(nextSet).findViewById<View>(R.id.leader) as TextView
            leader.text = getText(R.string.leader).toString() + ":" + getText(R.string.nothing)
        } else if (nextSet != -1) {//最大数登録したら終わり(indexが-1返される)
            //リーダー登録
            leaderArray[nextSet] = selectedMember
            val leader =
                groupListView.getChildAt(nextSet).findViewById<View>(R.id.leader) as TextView
            leader.text = getText(R.string.leader).toString() + ":" + selectedMember.name
            nextSet = leaderArray.indexOfFirst { it == null }
        }

        mbAdapter =
            SmallMBListAdapter(this, memberArray, leaderArray = leaderArray, showLeaderNo = true)
        memberListView.adapter = mbAdapter
    }

    //GroupArrayの作成
    private fun createGroupArray() {
        newGroupArray = ArrayList()
        for (i in 0 until groupListView.count) {
            val groupName = editGPAdapter!!.getGroupName(i)
            val memberNo = editGPAdapter!!.getMemberNo(i)
            newGroupArray.add(Group(i, groupName, "", memberNo))
        }
    }

    //キーボードによるレイアウト崩れを防ぐ
    private fun setKeyboardListener() {
        val activityRootView = findViewById<View>(R.id.custom_root_layout)
        val view = findViewById<View>(R.id.normal_kumiwake_button)
        activityRootView.viewTreeObserver.addOnGlobalLayoutListener(object :
            ViewTreeObserver.OnGlobalLayoutListener {
            private val r = Rect()

            override fun onGlobalLayout() {
                activityRootView.getWindowVisibleDisplayFrame(r)
                val heightDiff = activityRootView.rootView.height - r.height()
                if (heightDiff > screenHeight * 0.2) {
                    view.visibility = View.GONE
                } else {
                    view.visibility = View.VISIBLE
                }
            }
        })
    }
}